modules/src/core/ics05_port/context.rs

```diff
@@ -1,11 +1,68 @@
+use alloc::format;
+
 use crate::core::ics05_port::capabilities::Capability;
 use crate::core::ics05_port::error::Error;
 use crate::core::ics24_host::identifier::PortId;
+use crate::core::ics24_host::path::PortsPath;
+
+/// A context supplying all the necessary read-only dependencies for processing any information regarding a port.
+pub trait PortReader: CapabilityReader {
+    /// Module Id type that can be mapped to an ICS26 router callback module
+    type ModuleId;
+
+    /// Return the module_id along with the capability associated with a given port_id
+    fn lookup_module_by_port(
+        &self,
+        port_id: &PortId,
+    ) -> Result<(Self::ModuleId, Capability), Error>;
+
+    /// Check if the specified port_id is already bounded
+    fn is_bound(&self, port_id: PortId) -> bool {
+        self.get_capability(format!("{}", PortsPath(port_id)))
+            .is_ok()
+    }
+
+    /// Authenticate a capability key against a port_id by checking if the capability was previously
+    /// generated and bound to the specified port
+    fn authenticate(&self, port_id: PortId, capability: &Capability) -> bool {
+        self.authenticate_capability(format!("{}", PortsPath(port_id)), capability)
+            .is_ok()
+    }
+}
 
-// A context supplying all the necessary read-only dependencies for processing any information regarding a port.
-pub trait PortReader {
-    fn lookup_module_by_port(&self, port_id: &PortId) -> Result<Capability, Error>;
-    fn authenticate(&self, key: &Capability, port_id: &PortId) -> bool;
+pub trait PortKeeper: CapabilityKeeper + PortReader {
+    /// Binds to a port and returns the associated capability
+    fn bind_port(&mut self, port_id: PortId) -> Result<Capability, Error> {
+        if self.is_bound(port_id.clone()) {
+            Err(Error::port_already_bound(port_id))
+        } else {
+            self.new_capability(format!("{}", PortsPath(port_id)))
+        }
+    }
 }
 
-//  Result<Capability, Error>//return Ok(Capability::new());
+pub trait CapabilityKeeper {
+    /// Create a new capability with the given name.
+    /// Return an error if the capability was already taken.
+    fn new_capability(&mut self, name: impl AsRef<str>) -> Result<Capability, Error>;
```

Comment:
Is `AsRef<str>` really the best representation we can find for capabilities names? What are the operations on the capability name that are actually required? Perhaps we can define a trait for that and use `impl CapabilityName` instead?

Reply:
I just looked up the SDK impl and it seems capability names are expected to be non-empty strings (i.e. `strings.TrimSpace(name) != ""`) -> https://github.com/cosmos/cosmos-sdk/blob/master/x/capability/keeper/keeper.go#L230 , so I think it is better to use a newtype.

Reply:
Sweet!